File: src/leetCodeProblems/PriorityQueue/TopKFrequentElements347.java
package leetCodeProblems.PriorityQueue;

/**
 * LeetCode - https://leetcode.com/problems/top-k-frequent-elements/submissions/
 *
 * TimeComplexity - O(n) + O(klogk), where k is the count of unique frequencies in the array
 * SpaceComplexity - O(n)
 */

import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.PriorityQueue;

public class TopKFrequentElements347 {

    static class ElementFrequency {
        int element;
        int count;

        ElementFrequency(int element, int count) {
            this.element = element;
            this.count = count;

        }
    }

    static class ElementComparator implements Comparator<ElementFrequency> {

        public int compare(ElementFrequency elem1, ElementFrequency elem2) {
            return elem2.count - elem1.count;
        }
    }

    public int[] topKFrequent(int[] nums, int k) {

        HashMap<Integer, Integer> frequencyMap = new HashMap();
        PriorityQueue<ElementFrequency> frequencyPQ = new PriorityQueue<>(new ElementComparator());

        int[] output = new int[k];

        for (int num: nums) {
            frequencyMap.put(num,
                    frequencyMap.getOrDefault(num, 0) + 1);
        }

        for (int element: frequencyMap.keySet()) {
            ElementFrequency obj = new ElementFrequency(element, frequencyMap.get(element));
            frequencyPQ.add(obj);
        }

        for (int i=0; i <k; i++) {
            output[i] = frequencyPQ.poll().element;
        }

        return output;
    }

    public static void main(String[] args) {

        int[] input = {1,1,1,2,2,3};

        TopKFrequentElements347 obj = new TopKFrequentElements347();

        int[] output = obj.topKFrequent(input, 2);

        System.out.print(Arrays.toString(output));

    }
}

File: src/leetCodeProblems/PriorityQueue/TopKFrequentElements347.go
package main

/*
- LeetCode - https://leetcode.com/problems/top-k-frequent-words/submissions/
- Time - O(N) + O(N log K) = O(N log K)
- Space - O(N + K)
*/
import (
	"container/heap"
)

type Element struct {
	Num  int
	Freq int
}

type minHeap []Element

// Min Heap
func (h minHeap) Less(i, j int) bool {
	return h[i].Freq < h[j].Freq
}

func (h minHeap) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
}

func (h minHeap) Len() int {
	return len(h)
}

func (h *minHeap) Push(i interface{}) {
	*h = append(*h, i.(Element))
}

func (h *minHeap) Pop() interface{} {
	t := (*h)[len(*h)-1]
	*h = (*h)[:len(*h)-1]
	return t
}

func topKFrequent1(nums []int, k int) []int {
	m := make(map[int]int)
	h := new(minHeap)
	heap.Init(h)

	for _, v := range nums {
		m[v]++
	}

	for n, f := range m {
		if h.Len() >= k {
			e := heap.Pop(h).(Element)
			if e.Freq < f {
				heap.Push(h, Element{n, f})
			} else {
				heap.Push(h, e)
			}
		} else {
			heap.Push(h, Element{n, f})
		}
	}

	out := []int{}

	for i := 0; i < k; i++ {
		out = append(out, heap.Pop(h).(Element).Num)
	}
	return out
}
